python/src/main/python/drivers/runspv.py

```diff
@@ -0,0 +1,944 @@
+        # If the legacy worker is being used, give it the right permissions, stop it if already running, and get its
+        # working directory ready.
+        adb_check('shell pm grant com.graphicsfuzz.vkworker android.permission.READ_EXTERNAL_STORAGE')
+        adb_check('shell pm grant com.graphicsfuzz.vkworker android.permission.WRITE_EXTERNAL_STORAGE')
+        adb_can_fail('shell am force-stop ' + ANDROID_LEGACY_APP)
+        adb_can_fail('shell rm -rf ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR)
+        adb_check('shell mkdir -p ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR)
+    else:
+        # TODO: check that amber_ndk is under ANDROID_DEVICE_DIR
+        adb_can_fail('shell rm -rf ' + ANDROID_DEVICE_GRAPHICSFUZZ_DIR)
+        adb_check('shell mkdir -p ' + ANDROID_DEVICE_GRAPHICSFUZZ_DIR)
+
+    if wait_for_screen:
+        stay_awake_warning()
+        # We cannot reliably know if the screen is on, but this function definitely knows if it is
+        # off or locked. So we wait here while we definitely know there is an issue.
+        while is_screen_off_or_locked():
+            print('\nWARNING: The screen appears to be off or locked. Please unlock the device and '
+                  'ensure "Stay Awake" is enabled in developer settings.\n')
+            time.sleep(BUSY_WAIT_SLEEP_SLOW)
+
+
+################################################################################
+# Legacy worker: image test
+
+
+def run_image_android_legacy(vert: str, frag: str, json: str, output_dir: str, force: bool, skip_render: bool):
+    assert (os.path.isfile(vert))
+    assert (os.path.isfile(frag))
+    assert (os.path.isfile(json))
+
+    logfile = output_dir + os.sep + LOGFILE_NAME
+    statusfile = output_dir + os.sep + 'STATUS'
+
+    prepare_device(not force, True)
+
+    adb_check('push ' + vert + ' ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR + '/test.vert.spv')
+    adb_check('push ' + frag + ' ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR + '/test.frag.spv')
+    adb_check('push ' + json + ' ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR + '/test.json')
+
+    cmd = 'shell am start -n ' + ANDROID_LEGACY_APP + '/android.app.NativeActivity'
+    flags = '--num-render {}'.format(NUM_RENDER)
+    if skip_render:
+        flags += ' --skip-render'
+
+    # Pass command line args as Intent extra. Need to nest-quote, hence the "\'flags\'"
+    cmd += ' -e gfz "\'' + flags + '\'"'
+    adb_check(cmd)
+
+    # Busy wait
+    deadline = time.time() + TIMEOUT_APP
+
+    status = 'UNEXPECTED_ERROR'
+
+    # If the app never starts, status remains as UNEXPECTED_ERROR.
+    # Once the app starts, status becomes TIMEOUT.
+    # If we reach the end of the loop below, we break and the status is updated just before.
+
+    while time.time() < deadline:
+        time.sleep(BUSY_WAIT_SLEEP_SLOW)
+
+        # Don't pass here until app has started.
+        if status == 'UNEXPECTED_ERROR':
+            if adb_can_fail('shell test -f ' + ANDROID_SDCARD_GRAPHICSFUZZ_DIR + '/STARTED').returncode != 0:
```

Comment:
All adb_can_fail() commands may raise a subprocess.TimeoutExpired exception, which I guess we will want to catch in such busy-waiting loops.

Reply:
Yeah it is hard to know what do. If we get a timeout after 30 seconds running an adb command then something has gone very wrong. We could catch this and log an "unexpected error", so that the worker is likely to continue running. However, the worker is likely to keep returning unexpected errors from this point forward, polluting all future results. If we don't catch it, the exception will propagate to the glsl-to-spv worker (because we call runspv via Python), and the worker will likely terminate. This might be the right thing to do because the Android device is in a weird state and needs to be investigated, and all future tests are probably going to fail anyway.

Reply:
I vote to keep this